src/main/scala/Ch2.scala

```diff
@@ -0,0 +1,111 @@
+package swscala
+
+
+object Exercises1 {
+
+  val zeroToNine = (0 to 9)
+
+  // Problem 1
+  lazy val p1: Seq[(Int, Int)] = zeroToNine.flatMap(i => zeroToNine.map(j => (i, j)))
+    .filter{case (i, j) => i + 4 * j  > i * j}
+
+  // Problem 2
+  def p2(ss: Seq[String], bs: Seq[Boolean]): Seq[String] =
+    ss.zip(bs).filter(_._2).map(_._1)
+
+  // Problem 3
+  def p3(nums: Seq[Int]): Seq[(Int, Boolean)] =
+    nums.zip(nums.zip(nums.tail).map{case (former: Int, latter: Int) => former < latter})
+
+  // Problem 4
+  def p4[S, I](a: Seq[S], b: Seq[I]): Map[I, S] = b.zip(a).toMap
+
+  // Problem 5
+  def p5[S](a: Seq[S], b: Seq[Int]) = a.zip(b).sortBy(_._2).map(_._1)
+}
+
+object Exercises2 {
+
+  // Problem 1
+  def p1(purchasedItems: Seq[(String, Int)]): Map[String, Int] =
+    purchasedItems.groupBy(_._1).map{case (k, v) => (k, v.map(_._2).sum)}
+
+  // Problem 2
+  def p2(numsLists: Seq[List[Int]]): Seq[List[Int]] =
+    numsLists.map(_.sortBy(-_).take(3))
+
+  // Problem 3
+  def p3[I, J](a: Set[I], b: Set[J]): Set[(I, J)] =
+    a.flatMap(x => b.map(y => (x, y)))
+
+  // Problem 4
+  def p4[Person, Amount](daily: Seq[Map[Person, Amount]]): Map[Person, Seq[Amount]] =
+    daily.flatMap(_.toSeq).groupBy(_._1).map{case (k, v) => (k, v.map(_._2))}
+}
+
+object Exercises3 {
+
+
+  // Problem 1
+  def sumDigits(f: Int => Int)(num: Int): Int =
+    num.toString.map(d => f(d.asDigit)).sum
+
+  val sqSumDigits: Int => Int = sumDigits((x:Int) => x * x)
```

Comment:
I think the code will compile with no type annotations at all, actually.
```scala
val sqSumDigits= sumDigits(x => x * x)
```

Reply:
`val sqSumDigits = sumDigits((x:Int) => x * x)`
gave me this:
```
[info] Compiling 1 Scala source to /Users/joseph.kim/sw-scala/target/scala-2.12/classes ...
[error] /Users/joseph.kim/sw-scala/src/main/scala/Ch2.scala:53:30: missing argument list for method sumDigits in object Exercises3
[error] Unapplied methods are only converted to functions when a function type is expected.
[error] You can make this conversion explicit by writing `sumDigits _` or `sumDigits(_)(_)` instead of `sumDigits`.
[error]   val sqSumDigits = sumDigits((x:Int) => x * x)
[error]                              ^
[error] one error found
[error] (compile:compileIncremental) Compilation failed
[error] Total time: 0 s, completed Dec 12, 2017 11:33:12 PM
```
So then I tried `val sqSumDigits = sumDigits(x => x * x) _`
and it worked!